### PR TITLE
Combine Stunnel with Rsync

### PR DESCRIPTION
**Pull request by shubham-pampattiwar**
This PR does the following:
- Stunnel Configuration changes:
    - Removal of `foreground = yes` so that the stunnel process runs as daemon.
    - Addition of `output = /dev/stdout` for streaming of stunnel process logs as we are running the process in daemon mode.
    - Updating `accept = localhost:2222` in order to facilitate data transfer between rsync and stunnel over the lcoalhost interface.
- Transfer Image changes: These involve the addition of new packages/utilities used for embedding new workflow log in the container commands of rsync and stunnel.
- Accommodate Stunnel alongside Rsync in the `RunRsyncOperations` phase:
    - Modifying `getRsyncClientPodRequirements` to have destination IP for rsync process as `localhost`.
    - Modifying  `getRsyncClientPodTemplate` :
        - To add Stunnel Volumes consisting of stunnel configuration file, stunnel certs as well as the volume called `rsync-stunnel-ipc` (used for facilitating inter-process communication between rsync and stunnel).  
        - To add Stunnel VolumeMounts of stunnel certs, configs, `rsync-stunnel-ipc`. 
        - To add `rsync-stunnel-ipc` alongside the PV involved in the migration of the application.
        - To add bash scripts for rsync container command as well as the stunnel contaienr command.
        - To add stunnel container manifest consisting of all the above changes stunnel changes.

Enhancement doc reference: https://github.com/konveyor/enhancements/pull/18
Dependent PRs: https://github.com/konveyor/rsync-transfer/pull/4

### CODE REVIEW

File: pkg/controller/directvolumemigration/rsync.go [hunk 1]
```diff
@@ -1611,12 +1611,35 @@ func (req rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
...
+			MountPath: "/etc/stunnel/stunnel.conf",
+			SubPath:   "stunnel.conf",
+		},
+		{
+			Name:      "stunnel-certs",
+			MountPath: "/etc/stunnel/certs",
+		},
+		{
+			Name:      "rsync-stunnel-ipc",
+			MountPath: "/rsync-data",
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigration/rsync.go:1639]: should this be `/usr/share/rsync-data`?

File: pkg/controller/directvolumemigration/rsync.go [hunk 2]
```diff
@@ -1611,12 +1611,35 @@ func (req rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
...
-	volumeMounts = append(volumeMounts, corev1.VolumeMount{
+	rsyncVolumeMounts = append(rsyncVolumeMounts, corev1.VolumeMount{
 		Name:      req.pvInfo.name,
 		MountPath: fmt.Sprintf("/mnt/%s/%s", req.namespace, req.pvInfo.name),
 	})
+
+	// shared volumeMount for inter-process communication between rsync and stunnel
+	rsyncVolumeMounts = append(rsyncVolumeMounts, corev1.VolumeMount{
+		Name:      "rsync-stunnel-ipc",
+		MountPath: "/usr/share/rsync-data",
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigration/rsync.go:1624]: NIT: `rsync-data` has got a different ring to it than what you want may be make this mountPath `/usr/share/rsync-stunnel-mgmt` ? or something similar?

File: pkg/controller/directvolumemigration/rsync.go [hunk 3]
```diff
@@ -1684,15 +1798,17 @@ func (req rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
...
-	transferImage, err := cluster.GetRsyncTransferImage(t.Client)
-	if err != nil {
-		return req, liberr.Wrap(err)
-	}
+	//transferImage, err := cluster.GetRsyncTransferImage(t.Client)
+	//if err != nil {
+	//	return req, liberr.Wrap(err)
+	//}
+	// will update this once the rsync-trasnfer PR is merged
+	transferImage := "quay.io/spampatt/rsync-transfer:latest"
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigration/rsync.go:1811]: drop the comments? the transferImage variable on migrationcontroller can be used for testing?

**Review by alaypatel07 — CHANGES_REQUESTED** [review]
The changes look great for the first pass, I added a few NITs. Also, I realized we are missing the unit tests for `getRsyncClientContainerStatus` as a part of proving these changes are not introducing regression, would it be fair to ask adding unit tests covering `getRsyncClientContainerStatus`? 

P.S. @shubham-pampattiwar adding UTs might not be a big deal as it might initially seem. You can ping @pranavgaikwad and I for any sort of guidance you need.

File: pkg/controller/directvolumemigration/rsync.go [hunk 4]
```diff
@@ -1625,10 +1648,78 @@ func (req rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
...
+		},
+	})
+
 	rsyncCommand := []string{"rsync"}
 	rsyncCommand = append(rsyncCommand, req.rsyncOptions...)
 	rsyncCommand = append(rsyncCommand, fmt.Sprintf("/mnt/%s/%s/", req.namespace, req.pvInfo.name))
 	rsyncCommand = append(rsyncCommand, fmt.Sprintf("rsync://root@%s/%s", req.destIP, req.pvInfo.name))
+
+	rsyncCommandStr := strings.Join(rsyncCommand, " ")
+	log.Info(fmt.Sprintf("Rsync command executes: %s", rsyncCommandStr))
```
Comment (pranavgaikwad) [inline pkg/controller/directvolumemigration/rsync.go:1702]: this is logged further ahead. can drop this log

File: pkg/controller/directvolumemigration/rsync.go [hunk 5]
```diff
@@ -1645,22 +1736,45 @@ func (req rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
...
+		Command:                  stunnelContainerCommand,
+		Ports: []corev1.ContainerPort{
+			{
+				Name:          DirectVolumeMigrationStunnel,
+				Protocol:      corev1.ProtocolTCP,
+				ContainerPort: int32(2222),
+			},
+		},
+		VolumeMounts: stunnelVolumeMounts,
+		SecurityContext: &corev1.SecurityContext{
```
Comment (pranavgaikwad) [inline pkg/controller/directvolumemigration/rsync.go:1770]: Does the Stunnel container use same SecurityContext settings as Rsync container?

File: pkg/controller/directvolumemigration/stunnel.go [hunk 6]
```diff
@@ -60,13 +54,13 @@ metadata:
...
   stunnel.conf: |
-    foreground = yes
     pid =
     sslVersion = TLSv1.2
     client = yes
     syslog = no
+    output = /dev/stdout
     [rsync]
-    accept = {{ .StunnelPort }}
+    accept = localhost:2222
```
Comment (pranavgaikwad) [inline pkg/controller/directvolumemigration/stunnel.go:63]: any reason for not using `localhost:{{ .StunnelPort }}` ?

File: pkg/controller/directvolumemigration/rsync.go [hunk 7]
```diff
@@ -1645,22 +1736,45 @@ func (req rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
...
+			Privileged:             &isPrivileged,
+			RunAsUser:              &runAsUser,
+			ReadOnlyRootFilesystem: &trueBool,
```
Comment (pranavgaikwad) [inline pkg/controller/directvolumemigration/rsync.go:1773]: What happened to the `dropCapabilities` fields present in the previous template?

File: pkg/controller/directvolumemigration/rsync.go [hunk 8]
```diff
@@ -1645,22 +1736,45 @@ func (req rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
...
+		},
+		Resources: req.resourceReq,
```
Comment (pranavgaikwad) [inline pkg/controller/directvolumemigration/rsync.go:1775]: Using same resources as Rsync container? `resourceReq` was specifically for Rsync container

**Review by pranavgaikwad — CHANGES_REQUESTED** [review]
Added some questions in general. Needs some work around separating Rsync and Stunnel containers resource requirements.

File: pkg/controller/directvolumemigration/rsync.go (hunk 1 above)
Comment (alaypatel07) [reply pkg/controller/directvolumemigration/rsync.go:1639]: w.r.t functionality it does not matter but good to keep that same. That way we don't have to look at the YAML everytime we are debugging this

File: pkg/controller/directvolumemigration/rsync.go [hunk 9]
```diff
@@ -1595,8 +1595,10 @@ type rsyncClientPodRequirements struct {
 	password string
 	// privileged whether Rsync Pod will run privileged
 	privileged bool
-	// resourceReq resource requirements for the Pod
-	resourceReq corev1.ResourceRequirements
+	// rsyncResourceReq resource requirements for the rsync container
+	rsyncResourceReq corev1.ResourceRequirements
+	// stunnelResourceReq resource requirements for the rsync container
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigration/rsync.go:1600]: stunnelResourceReq resource requirements for the ***_stunnel_*** container

File: pkg/controller/directvolumemigration/rsync.go (hunk 1 above)
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigration/rsync.go]: Done updated it to `/usr/share/rsync-stunnel-mgmt` for uniformity.

File: pkg/controller/directvolumemigration/rsync.go (hunk 2 above)
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigration/rsync.go]: Used `rsync-stunnel-mgmt`

File: pkg/controller/directvolumemigration/rsync.go [hunk 10]
```diff
@@ -1625,10 +1648,78 @@ func (req rsyncClientPodRequirements) getRsyncClientPodTemplate() corev1.Pod {
...
 	rsyncCommand = append(rsyncCommand, req.rsyncOptions...)
 	rsyncCommand = append(rsyncCommand, fmt.Sprintf("/mnt/%s/%s/", req.namespace, req.pvInfo.name))
 	rsyncCommand = append(rsyncCommand, fmt.Sprintf("rsync://root@%s/%s", req.destIP, req.pvInfo.name))
+
+	rsyncCommandStr := strings.Join(rsyncCommand, " ")
+	log.Info(fmt.Sprintf("Rsync command executes: %s", rsyncCommandStr))
+	rsyncContainerCommand := []string{
+		"/bin/bash",
+		"-c",
+		"trap \"touch /usr/share/rsync-data/rsync-client-container-done\" exit $rc; timeout=600; SECONDS=0; while [ $SECONDS -lt $timeout ]; do nc -z localhost 2222; if [ $? -eq 0 ]; then " + rsyncCommandStr + "; rc=$?; break; fi; done;",
```
[the remark this answers is not in the document]
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigration/rsync.go]: Done, updated the PR

File: pkg/controller/directvolumemigration/rsync.go (hunk 5 above)
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigration/rsync.go:1775]: Yes, updated the PR and also added Drop capabilities. Reference PR for SCC: https://github.com/konveyor/mig-controller/pull/1001/files

File: pkg/controller/directvolumemigration/stunnel.go (hunk 6 above)
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigration/stunnel.go]: Done, updated the PR, had missed it, thanks :)

File: pkg/controller/directvolumemigration/rsync.go (hunk 8 above)
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigration/rsync.go]: Added changes, updated the PR.

**Comment by shubham-pampattiwar** [comment]
@alaypatel07 @pranavgaikwad Please take another look at this PR, I have updated it with the missing stunnel requirements, minor fixes as well as added unit tests for DVMP's `getRsyncClientContainerStatus`

File: pkg/controller/directvolumemigration/rsync.go [hunk 11]
```diff
@@ -1604,8 +1604,10 @@ type rsyncClientPodRequirements struct {
 	password string
 	// privileged whether Rsync Pod will run privileged
 	privileged bool
-	// resourceReq resource requirements for the Pod
-	resourceReq corev1.ResourceRequirements
+	// rsyncResourceReq resource requirements for the rsync container
+	rsyncResourceReq corev1.ResourceRequirements
```
Comment (pranavgaikwad) [inline pkg/controller/directvolumemigration/rsync.go:1608]: +1

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller.go [hunk 12]
```diff
@@ -195,7 +199,7 @@ func (r *RsyncPodProgressTask) Run() error {
 		if err != nil {
 			return liberr.Wrap(err)
 		}
-		rsyncPodStatus, err := r.getRsyncClientContainerStatus(pod)
+		rsyncPodStatus, err := r.getRsyncClientContainerStatus(pod, r)
```
Comment (pranavgaikwad) [inline pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller.go:202]: @shubham-pampattiwar Injecting the dependency here (duplication of `r`) feels like we need to change the design of struct `RsyncPodProgressTask`. But for now, this looks OK to me.

**Review by pranavgaikwad — APPROVED** [review]

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go [hunk 13]
```diff
@@ -336,3 +340,565 @@ func Test_getLastMatch(t *testing.T) {
...
+							},
+						},
+					},
+					Status: kapi.PodStatus{
+						Phase: kapi.PodSucceeded,
+						ContainerStatuses: []kapi.ContainerStatus{
+							{
+								Name: "stunnel",
+								State: kapi.ContainerState{
+									Running: &kapi.ContainerStateRunning{
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go:590]: if the pod succeeded this should not be running right? it should also be terminated?

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go [hunk 14]
```diff
@@ -336,3 +340,565 @@ func Test_getLastMatch(t *testing.T) {
...
+								Name: "rsync-client",
+							},
+						},
+					},
+					Status: kapi.PodStatus{
+						ContainerStatuses: []kapi.ContainerStatus{
+							{
+								Name: "stunnel",
+								State: kapi.ContainerState{
+									Running: &kapi.ContainerStateRunning{
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go:756]: same here, if the other container failed, this would not be running right?

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go [hunk 15]
```diff
@@ -336,3 +340,565 @@ func Test_getLastMatch(t *testing.T) {
...
+				PodPhase:                    kapi.PodFailed,
+				LogMessage:                  "",
+				LastObservedProgressPercent: "",
+				LastObservedTransferRate:    "",
+				ExitCode:                    &one,
+			},
+			wantErr: false,
+		},
+		{
+			name: "Given a non ready rsync container and refPod in pending state, appropriate container status is returned",
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go:790]: drop "appropriate container status is returned" from all TC names :)

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go [hunk 16]
```diff
@@ -336,3 +340,565 @@ func Test_getLastMatch(t *testing.T) {
...
+							},
+						},
+					},
+					Status: kapi.PodStatus{
+						Phase: kapi.PodPending,
+						ContainerStatuses: []kapi.ContainerStatus{
+							{
+								Name: "stunnel",
+								State: kapi.ContainerState{
+									Running: &kapi.ContainerStateRunning{
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go:839]: cant be running when in `Pending`

**Review by alaypatel07 — APPROVED** [review]
Added small nits to this, overall looks great, VISIACK

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go (hunk 13 above)
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go]: Done, updated the PR

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go (hunk 14 above)
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go]: Done, updated the PR

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go (hunk 16 above)
Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller_test.go]: Done, updated the PR

**Review by alaypatel07 — APPROVED** [review]
VISIACK

File: pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller.go [hunk 17]
```diff
@@ -347,7 +347,8 @@ func (r *RsyncPodProgressTask) getRsyncClientContainerStatus(podRef *kapi.Pod, p
 		LogMessage:        podRef.Status.Message,
 	}
 	var containerStatus *kapi.ContainerStatus
-	for _, c := range podRef.Status.ContainerStatuses {
+	for i := range podRef.Status.ContainerStatuses {
+		c := podRef.Status.ContainerStatuses[i]
 		if c.Name == RsyncContainerName {
```
Comment (alaypatel07) [inline pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller.go:352]: Why is this change needed?

Comment (alaypatel07) [reply pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller.go:352]: I mean two lines above

Comment (shubham-pampattiwar) [reply pkg/controller/directvolumemigrationprogress/directvolumemigrationprogress_controller.go:352]: So this piece of code was taking the reference of the index variable and not the actual status reference (was not observed earlier because there was only on container in the Pod). This was identified by @pranavgaikwad while he was testing out the PR.
